tooling/greybox_fuzzer/src/mutation/mod.rs

```diff
@@ -0,0 +1,769 @@
+mod dictionary;
+mod field;
+mod int;
+mod string;
+use num_traits::Zero;
+pub struct InputMutator {
+    abi: Abi,
+    weight_tree: NodeWeight,
+    full_dictionary: FullDictionary,
+}
+
+const MUTATION_LOG_MIN: u32 = 0;
+const MUTATION_LOG_MAX: u32 = 5;
+/// NodeWeight determines the probability of mutating a particular object
+/// It represent the relative weight of this InputValue and its children
+#[derive(Clone, Debug)]
+pub struct NodeWeight {
+    start: u32,
+    end: u32,
+    subnodes: Option<Vec<NodeWeight>>,
+}
+impl NodeWeight {
+    pub fn get_weight(&self) -> u32 {
+        self.end - self.start
+    }
+    pub fn calculate_offsets(&mut self, start_offset: u32) {
+        assert!(self.start == 0, "Only calculate offsets after initial computation");
+        self.start += start_offset;
+        self.end += start_offset;
+        let mut current_update = start_offset;
+        if self.subnodes.is_some() {
+            let subnode_weights = self.subnodes.as_mut().unwrap();
+            for subnode_weight in subnode_weights.iter_mut() {
+                let weight_update = subnode_weight.get_weight();
+                subnode_weight.calculate_offsets(current_update);
+                current_update += weight_update;
+            }
+        }
+    }
+}
+
+enum UnbalancedSplice {
+    FirstTestCase,
+    SecondTestCase,
+}
+
+struct UnbalancedSpliceConfiguration {
+    first_testcase_weight: usize,
+    #[allow(unused)]
+    second_testcase_weight: usize,
+    total_weight: usize,
+}
+
+impl UnbalancedSpliceConfiguration {
+    #[allow(unused)]
+    pub fn new(first_testcase_weight: usize, second_testcase_weight: usize) -> Self {
+        let total_weight = first_testcase_weight + second_testcase_weight;
+        Self { first_testcase_weight, second_testcase_weight, total_weight }
+    }
+
+    /// Select a mutation according to weights
+    pub fn select(&self, prng: &mut XorShiftRng) -> UnbalancedSplice {
+        let selector = prng.gen_range(0..self.total_weight);
+        if selector < self.first_testcase_weight {
+            return UnbalancedSplice::FirstTestCase;
+        }
+        UnbalancedSplice::SecondTestCase
+    }
+}
+
+const BASIC_UNBALANCED_SLICE_CONFIGURATION: UnbalancedSpliceConfiguration =
+    UnbalancedSpliceConfiguration {
+        first_testcase_weight: 8,
+        second_testcase_weight: 2,
+        total_weight: 8 + 2,
+    };
+const BOOL_WEIGHT: u32 = 1u32;
+const FIELD_WEIGHT: u32 = 8u32;
+const STRING_WEIGHT_MULTIPLIER: u32 = 4;
+impl InputMutator {
+    /// Create a new input mutator given the ABI and a hashset of field elements used in the circuit (for the dictionary)
+    pub fn new(abi: &Abi, original_dictionary: &HashSet<FieldElement>) -> Self {
+        let mut weight_tree = Self::count_all_input_weights(abi);
+        weight_tree.calculate_offsets(0);
+        Self {
+            abi: abi.clone(),
+            weight_tree,
+            full_dictionary: FullDictionary::new(original_dictionary),
```

Comment:
`FullDictionary::new(original_dictionary)` could be extracted to `let full_dictionary = ...` and it will reduce the number of lines here.